x-pack/plugins/security/server/lib/authorization/has_privileges.js

```diff
@@ -0,0 +1,150 @@
+  if (!hasPrivileges[getVersionPrivilege(kibanaVersion)] && hasPrivileges[getLoginPrivilege()]) {
+    throw new Error('Multiple versions of Kibana are running against the same Elasticsearch cluster, unable to authorize user.');
+  }
+
+  return {
+    username: privilegeCheck.username,
+    hasAllRequested: privilegeCheck.has_all_requested,
+    privileges: hasPrivileges
+  };
+};
+
+const hasLegacyPrivileges = async (
+  savedObjectTypes,
+  deprecationLogger,
+  callWithRequest,
+  request,
+  kibanaVersion,
+  application,
+  kibanaIndex,
+  privileges
+) => {
+  const privilegeCheck = await callWithRequest(request, 'shield.hasPrivileges', {
+    body: {
+      index: [{
+        names: [ kibanaIndex ],
+        privileges: ['read', 'index']
+      }]
+    }
+  });
+
+  const createPrivileges = (cb) => {
+    return privileges.reduce((acc, name) => {
+      acc[name] = cb(name);
+      return acc;
+    }, {});
+  };
+
+  const logDeprecation = () => {
+    deprecationLogger(
+      `Relying on implicit privileges determined from the index privileges is deprecated and will be removed in the next major version`
+    );
+  };
+
+  // if they have the index privilege, then we grant them all actions
+  if (privilegeCheck.index[kibanaIndex].index) {
+    logDeprecation();
+    const implicitPrivileges = createPrivileges(() => true);
+    return {
+      username: privilegeCheck.username,
+      hasAllRequested: true,
+      privileges: implicitPrivileges
+    };
+  }
+
+  // if they have the read privilege, then we only grant them the read actions
+  if (privilegeCheck.index[kibanaIndex].read) {
+    logDeprecation();
+    const privilegeMap = buildPrivilegeMap(savedObjectTypes, application, kibanaVersion);
+    const implicitPrivileges = createPrivileges(name => privilegeMap.read.actions.includes(name));
+
+    return {
+      username: privilegeCheck.username,
+      hasAllRequested: Object.values(implicitPrivileges).every(x => x),
+      privileges: implicitPrivileges,
+    };
+  }
+
+  return {
+    username: privilegeCheck.username,
+    hasAllRequested: false,
+    privileges: createPrivileges(() => false)
+  };
+};
+
+export function hasPrivilegesWithServer(server) {
+  const callWithRequest = getClient(server).callWithRequest;
+
+  const config = server.config();
+  const kibanaVersion = config.get('pkg.version');
+  const application = config.get('xpack.security.rbac.application');
+  const kibanaIndex = config.get('kibana.index');
+  const savedObjectTypes = server.savedObjects.types;
+  const deprecationLogger = (msg) => server.log(['warning', 'deprecated', 'security'], msg);
+
+  return function hasPrivilegesWithRequest(request) {
+    return async function hasPrivileges(privileges) {
+      const loginPrivilege = getLoginPrivilege();
+      const versionPrivilege = getVersionPrivilege(kibanaVersion);
+
+      const allPrivileges = [versionPrivilege, loginPrivilege, ...privileges];
```

Comment:
I'm curious about the `versionPrivilege`, not sure I get what it's for, can you explain it to me?

Reply:
The `versionPrivilege` is a safeguard to ensure that another version of Kibana didn't come online and overwrite the privilege definitions. The actions within each privilege have the potential to change drastically between versions, and we want to ensure that we are checking against a known set of privileges.

This is important because a different version of Kibana could introduce additional actions within an existing privilege, which would grant additional access to a user that they perhaps didn't have before.

Reply:
Do we always append application privileges for each version of Kibana? So if a user installs version x.0, they'll get appPrivs for x.0, and when we check for appPrivs we ensure they have the x.0 to be able to say "they at least have those privs". Then, when they install x.1 the appPrivs for x.1 will be added to the list and if for a short time x.0 is still operational somewhere they will see the appPriv for x.0 still. But once x.1 starts doing hasPrivs checks they will get a confirmation that the role also has privs for x.1?

Reply:
Feels to me like we would just put something like that in the application or resource... `"application": "kibana/x.0"` or something, then they're totally separate, but maybe we don't want to have to clean those up...